transformations/synonym_substitution/transformation.py

```diff
@@ -0,0 +1,98 @@
+import random
+import stanza
+
+stanza.download("en")
+
+import nltk
+
+nltk.download("wordnet")
+
+from nltk.corpus import wordnet
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+import re
+
+"""
+Base Class for implementing the different input transformations a generation should be robust against.
+"""
+
+
+def untokenize(words):
+    """
+    Untokenizing a text undoes the tokenizing operation, restoring
+    punctuation and spaces to the places that people expect them to be.
+    Ideally, `untokenize(tokenize(text))` should be identical to `text`,
+    except for line breaks.
+    ref: https://github.com/commonsense/metanl/blob/master/metanl/token_utils.py#L28
+    """
+    text = " ".join(words)
+    step1 = text.replace("`` ", '"').replace(" ''", '"').replace(". . .", "...")
+    step2 = step1.replace(" ( ", " (").replace(" ) ", ") ")
+    step3 = re.sub(r' ([.,:;?!%]+)([ \'"`])', r"\1\2", step2)
+    step4 = re.sub(r" ([.,:;?!%]+)$", r"\1", step3)
+    step5 = step4.replace(" '", "'").replace(" n't", "n't").replace("can not", "cannot")
+    step6 = step5.replace(" ` ", " '")
+    return step6.strip()
+
+
+def synonym_substitution(text, stanza_pipeline, seed=42, prob=0.5, max_outputs=1):
+    random.seed(seed)
+    upos_wn_dict = {
+        "VERB": "v",
+        "NOUN": "n",
+        "ADV": "r",
+        "ADJ": "s",
+    }
+
+    doc = stanza_pipeline(text)
```

Comment:
I would suggest you to use spacy since its already being used everywhere in NL-Augmenter unless stanza is required for anything unique.

Reply:
Updated. Thanks for the suggestion!